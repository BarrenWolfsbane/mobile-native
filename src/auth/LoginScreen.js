import React, {
  Component
} from 'react';

import { StackActions, NavigationActions } from 'react-navigation';
import FastImage from 'react-native-fast-image';

import {
  StyleSheet,
  ScrollView,
  View,
  KeyboardAvoidingView,
  Button,
  Keyboard,
  Animated,
  Platform,
} from 'react-native';
import * as Animatable from 'react-native-animatable';

import LoginForm from './LoginForm';
import VideoBackground from '../common/components/VideoBackground';
import { CommonStyle } from '../styles/Common';
import { ComponentsStyle } from '../styles/Components';
import logService from '../common/services/log.service';
import featuresService from '../common/services/features.service';

const LOGO_HEIGHT = 100;
const LOGO_HEIGHT_SMALL = 50;

/**
 * Login screen
 */
export default class LoginScreen extends Component {

  /**
   * Disable navigation bar
   */
  static navigationOptions = {
    header: null
  }

  constructor(props) {
    super(props);

    this.logoHeight = new Animated.Value(LOGO_HEIGHT);
  }

  componentWillMount () {
    this.keyboardWillShowSub = Keyboard.addListener('keyboardWillShow', this.keyboardWillShow);
    this.keyboardWillHideSub = Keyboard.addListener('keyboardWillHide', this.keyboardWillHide);
  }

  componentWillUnmount() {
    this.keyboardWillShowSub.remove();
    this.keyboardWillHideSub.remove();
  }

  keyboardWillShow = (event) => {
    Animated.timing(this.logoHeight, {
      duration: event.duration,
      toValue: LOGO_HEIGHT_SMALL,
    }).start();
  };

  keyboardWillHide = (event) => {
    Animated.timing(this.logoHeight, {
      duration: event.duration,
      toValue: LOGO_HEIGHT,
    }).start();
  };

  /**
   * Render
   */
  render() {
    const resizeMode = 'center';

    return (
      <KeyboardAvoidingView style={CommonStyle.flexContainer} behavior={ Platform.OS == 'ios' ? 'padding' : null }>
        <VideoBackground />
        <View style={[CommonStyle.flexContainerCenter, CommonStyle.padding2x]}>
          <Animatable.View animation="bounceIn">
            <Animated.Image
              resizeMode="contain"
              style={[styles.logo, { height: this.logoHeight }]}
              source={require('../assets/logos/logo-white.png')}
            />
          </Animatable.View>
          <Animatable.View animation="fadeInUp">
            <LoginForm
              onLogin={() => this.login()}
              onRegister={this.onPressRegister}
              onForgot={this.onPressForgot}
            />
          </Animatable.View>
        </View>
      </KeyboardAvoidingView>
    );
  }

  /**
   * On press forgot
   */
  onPressForgot = () => {
    this.props.navigation.push('Forgot');
  }

  /**
   * On press register
   */
  onPressRegister = () => {
    if (featuresService.has('on-boarding')) {
      this.props.navigation.push('RegisterNew');
    } else {
      this.props.navigation.push('Register');
    }
  }

  /**
   * On login successful
   */
  login() {
    logService.info('user logged in');
  }
}

const styles = StyleSheet.create({
  logo: {
    width: 200,
    height: 84,
    marginBottom: 30,
    alignSelf: 'center',
  },
});