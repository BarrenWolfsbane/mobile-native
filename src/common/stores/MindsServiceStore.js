import mindsService from "../services/minds.service";


export default class MindsServiceStore {
  loading = false;
  promise;

  async getSettings() {
    if (this.loading) {
      const result = await this.promise;
      this.loading = false;
      return result;
    }

    this.loading = true;

    this.promise = mindsService.getSettings();

    return await this.promise;
  }
}
