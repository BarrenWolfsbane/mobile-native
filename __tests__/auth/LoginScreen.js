import 'react-native';
import React from 'react';
import { Text, TouchableOpacity } from "react-native";
import { shallow } from 'enzyme';
import featuresService from '../../src/common/services/features.service';

import LoginScreen from '../../src/auth/LoginScreen';

jest.mock('../../src/auth/AuthService');
jest.mock('../../src/common/stores/MindsServiceStore');

jest.mock('../../src/auth/LoginForm', () => 'LoginForm');
jest.mock('../../src/auth/ForgotPassword', () => 'ForgotPassword');
jest.mock('../../src/common/components/VideoBackground', () => 'VideoBackground');

// Note: test renderer must be required after react-native.
import renderer from 'react-test-renderer';
import MindsServiceStore from '../../src/common/stores/MindsServiceStore'

describe('LoginScreen component', () => {

  beforeEach(() => {
    featuresService.features = {'homepage-december-2019': false};
  });

  it('should renders correctly', () => {
    const mindsServiceStore = new MindsServiceStore();
    const loginScreen = renderer.create(
      <LoginScreen mindsServiceStore={mindsServiceStore} />
    ).toJSON();
    expect(loginScreen).toMatchSnapshot();
  });

  it('should shows login form component', async () => {
    const mindsServiceStore = new MindsServiceStore();
    const wrapper = shallow(
      <LoginScreen hopFeatures={true} mindsServiceStore={mindsServiceStore} />
    );

    // search login form
    let render = wrapper.dive();
    const loginForms = render.find('LoginForm');

    // should contain 1 login form
    expect(loginForms.length).toBe(1);
  });

  it('should shows forgot password component if the user press the button', async () => {
    const mindsServiceStore = new MindsServiceStore();

    const navigation = {
      push: jest.fn()
    };

    const wrapper = shallow(
      <LoginScreen navigation={navigation} hopFeatures={true} mindsServiceStore={mindsServiceStore} />
    );

    // search login form
    let render = wrapper.dive();
    const loginForms = render.find('LoginForm');

    // should contain 1 login form
    expect(loginForms.length).toBe(1);

    // simulate forgot password component button press
    loginForms.at(0).props().onForgot();

    expect(navigation.push).toBeCalled();
  });
});
