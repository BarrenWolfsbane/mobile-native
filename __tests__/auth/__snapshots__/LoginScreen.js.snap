// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen component should renders correctly 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "flex": 1,
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <VideoBackground />
  <View
    style={
      Array [
        Object {
          "flex": 1,
          "justifyContent": "center",
        },
        Object {
          "padding": 10,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "opacity": 0,
          "transform": Array [
            Object {
              "scale": 0.3,
            },
          ],
        }
      }
    >
      <Image
        resizeMode="contain"
        source={
          Object {
            "testUri": "../../../src/assets/logos/logo-white.png",
          }
        }
        style={
          Object {
            "alignSelf": "center",
            "height": 100,
            "marginBottom": 30,
            "width": 200,
          }
        }
      />
    </View>
    <View
      style={
        Object {
          "opacity": 0,
          "transform": Array [
            Object {
              "translateY": 100,
            },
          ],
        }
      }
    >
      <LoginForm
        onForgot={[Function]}
        onLogin={[Function]}
        onRegister={[Function]}
      />
    </View>
  </View>
</View>
`;
